[data views] composite runtime fields in api layer

mattkime opened the pull request:
## Summary

Add composite runtime fields to the data view api layer. Broken out of https://github.com/elastic/kibana/pull/110223

Part of phase 3 of https://github.com/elastic/kibana-team/issues/307 

Composite runtime field ES docs - https://www.elastic.co/guide/en/elasticsearch/reference/current/runtime-examples.html#runtime-examples-grok-composite

---

To test composite runtime field, provide data view id  - 

```curl -u elastic -X POST "localhost:5601/api/data_views/data_view/${id}/runtime_field" -H 'kbn-xsrf: true' -H 'Content-Type: application/json' -d @"composite_runtime.txt"```

composite_runtime.txt content - 
```
{
  "name": "composite_runtime",
  "runtimeField": {
    "type": "composite",
    "script": {
      "source": "emit(\"a\",\"a\"); emit(\"b\",\"b\")"
    },
    "fields": {
      "a": {
        "type": "keyword"
      },
      "b": {
        "type": "keyword"
      }
    }
  }
}

```

and delete 
```
curl -u elastic -X DELETE "localhost:5601/api/data_views/data_view/${id}/runtime_field/composite_runtime" -H 'kbn-xsrf: true' -H 'Content-Type: application/json'
```

---

Testing
- Use API to submit composite runtime field
- Verify fields are in data view management. 
- Verify they cannot be deleted
- Verify they can be edited like mapped fields - can't change name or type, no set script ability
- Verify regular runtime fields work as expected
- Go to discover and verify that runtime fields work for search and query as expected
- Remove composite runtime field via API
- verify that all child fields have been removed.

mattkime commented on test/api_integration/apis/index_patterns/runtime_fields_crud/get_runtime_field/errors.ts:
```diff
@@ -68,16 +68,6 @@ export default function ({ getService }: FtrProviderContext) {
           );
         });
-
-        it('returns error when attempting to fetch a field which is not a runtime field', async () => {
```
I think this test is basically the same as the one on line 54.

mattkime commented on test/api_integration/apis/index_patterns/runtime_fields_crud/delete_runtime_field/errors.ts:
```diff
@@ -59,16 +59,6 @@ export default function ({ getService }: FtrProviderContext) {
           expect(response1.status).to.be(404);
         });
 
-        it('returns error when attempting to delete a field which is not a runtime field', async () => {
```
Same as test above, I don't see a need for a separate error code when removing a non runtime field

elasticmachine commented:
Pinging @elastic/kibana-app-services (Team:AppServicesSv)

flash1293 commented:
@mattkime The description of this PR doesn't really explain what's going on and the linked PR is pretty bare as well. Could you elaborate a bit on the context of these changes and why they are necessary? I want to make sure the changes done won't break anything else in VisEditors code as it seems like they are breaking in some ways.

mattkime commented:
@flash1293 Sorry for that, I've updated the description with additional context.

flash1293 reviewed (approved):
VisEditors changes LGTM

jbudz reviewed (approved):
limits.yml LGTM

mattkime commented:
@elasticmachine merge upstream

qn895 commented on x-pack/plugins/data_visualizer/public/application/index_data_visualizer/components/full_time_range_selector/full_time_range_selector_service.ts:
```diff
@@ -27,8 +27,7 @@ export async function setFullTimeRange(
 ): Promise<GetTimeFieldRangeResponse> {
-  const runtimeMappings = indexPattern.getComputedFields()
-    .runtimeFields as estypes.MappingRuntimeFields;
+  const runtimeMappings = indexPattern.getRuntimeMappings() as estypes.MappingRuntimeFields;
```
Since `getRuntimeMappings()` is already typed, we can potential remove this type cast here?

mattkime replied:
Nice catch, fixed!

stratoula reviewed (commented):
@mattkime we also use the `getComputedFields().runtimeFields` in TSVB. Do we want to replace it there too?

mattkime commented:
@elasticmachine merge upstream

mattkime commented:
@stratoula That sounds like a cleanup task to me.

Dosant commented:
@mattkime, see: https://github.com/elastic/kibana/issues/125397, not 100% sure if this is coming from this branch or from main. So created a separate issue

Dosant commented on src/plugins/data_views/server/rest_api_routes/runtime_fields/create_runtime_field.ts:
```diff
@@ -49,14 +49,11 @@ export const createRuntimeField = async ({
     throw new Error(`Field [name = ${name}] already exists.`);
```
Seems like this check doesn't work for composite fields

mattkime replied:
Fixed and added integration test to verify

Dosant commented on src/plugins/data_views/common/fields/data_view_field.ts:
```diff
@@ -108,13 +108,29 @@ export class DataViewField implements IFieldType {
+    if (this.isRuntimeField) {
+      let type: RuntimeType = this.runtimeField?.type!;
+      if (this.isRuntimeCompositeSubField) {
+        const [, subFieldName] = this.name.split('.');
```
I can currently have a composite field name `composite_name.a` and inside `emit('a','a');emit('b','b')` 

I think this `.split` check won't work currently in this case

Dosant replied:
Just for illustration the scenario where this check would work incorrectly: 


![Screen Shot 2022-02-11 at 17 21 37](https://user-images.githubusercontent.com/7784120/153628781-9e3484d6-ba59-4d53-8536-57a25bf602e9.png)



maybe the best fix would be for composite runtime field have it' original name available at `this.runtimeField.name`

mattkime replied:
Hm, its pretty clear that a bunch of this code needs to be rethought

mattkime replied:
Now type is set on a runtime field when the field list is created or updated.

qn895 commented on src/plugins/data_views/server/rest_api_routes/runtime_fields/get_runtime_field.ts:
```diff
@@ -43,17 +43,28 @@ export const getRuntimeField = async ({
-  if (!field.runtimeField) {
-    throw new Error('Only runtime fields can be retrieved.');
+  // Access the data view fields created for the runtime field
+  let dataViewFields: DataViewField[];
```
While the provided test case worked great, I tested with `params` in the payload and got this error:
```
{"statusCode":400,"error":"Bad Request","message":"[request body.runtimeField.script.params]: definition for this key is missing"}%                                                                                        
```

```
{
  "name": "composite_params",
  "runtimeField": {
    "type": "composite",
    "script": {
      "source": """emit("a","a"); emit("b","b");emit("c",params['c'])""",
      "params": {
        "c": "c"
      }
    },
    "fields": {
      "a": {
        "type": "keyword"
      },
      "b": {
        "type": "keyword"
      },
      "c": {
        "type": "keyword"
      }
    }
  }
}
```


Since this is a valid ES request, do we have any plans to support it in the future?

mattkime replied:
>Since this is a valid ES request, do we have any plans to support it in the future?

I don't believe so, although I'm interested to hear arguments for it. Its unclear to me how this is useful, in the context of data views, over providing the param value directly in the `source`

qn895 replied:
Thanks for the context. I don't have any arguments for it, just curious if it's in our future plan to match with what ES supports but that's out of scope of this PR. I'm happy to resolve this for now.

stephmilovic commented on x-pack/plugins/security_solution/public/common/components/inspect/modal.tsx:
```diff
@@ -153,22 +135,7 @@ export const ModalInspectQuery = ({
         <span data-test-subj="index-pattern-description">
-          <p>{formatIndexPatternRequested(inspectRequests[0]?.index ?? [])}</p>
-
-          {!isSourcererPattern && (
```
hey i just put this in! https://github.com/elastic/kibana/pull/125007

What gives?

mattkime replied:
I must have done a crap job of pulling changes from the other branch. sorry about that! I'll restore your changes and try adding back in the intended set of changes

mattkime replied:
@stephmilovic Take another look! I think I removed all the mistaken changes

stephmilovic commented on x-pack/plugins/security_solution/public/common/components/inspect/index.tsx:
```diff
@@ -138,23 +129,21 @@ const InspectButtonComponent: React.FC<InspectButtonProps> = ({
           onClick={handleClick}
         />
       )}
-      {isShowingModal && request !== null && response !== null && (
```
reverting all my good work? https://github.com/elastic/kibana/pull/125007

stephmilovic commented on x-pack/plugins/security_solution/public/common/components/exceptions/viewer/helpers.tsx:
```diff
@@ -127,38 +125,7 @@ export const getDescriptionListContent = (
   return details.reduce<DescriptionListItem[]>((acc, { value, title }) => {
     if (value != null && value.trim() !== '') {
-      const valueElement = (
-        <EuiToolTip content={value} anchorClassName="eventFiltersDescriptionListDescription">
```
this just went in 2 days ago: https://github.com/elastic/kibana/pull/125145

I think you maybe need to redo your merge with main and resolve these conflicts, you're deleting a lot of intentional code

qn895 commented on x-pack/plugins/transform/common/api_schemas/common.ts:
```diff
@@ -58,17 +56,27 @@ export interface CommonResponseStatusSchema {
-        }
-      },
-    }
+  schema.recordOf(
```
I don't think this change is needed isRuntimeField() is already be supporting 'composite' if it's listed in data.common's `RUNTIME_FIELD_TYPES`. Since users can create transform through API, we have to support more cases than what Kibana currently supports.

weltenwort commented on x-pack/plugins/infra/common/log_sources/resolved_log_source_configuration.ts:
```diff
@@ -119,7 +119,7 @@ const resolveRuntimeMappings = (indexPattern: DataView): estypes.MappingRuntimeF
               script: {
                 lang: 'painless', // required in the es types
-                source: runtimeFieldSpec.script.source,
+                source: (runtimeFieldSpec.script as estypes.InlineScript).source,
```
Since `runtimeFields` now corresponds to the mapping due to the changes in `data_view.ts`, no conversion is necessary anymore. This whole function should just be replaced with

```typescript
const resolveRuntimeMappings = (indexPattern: DataView): estypes.MappingRuntimeFields => {
  return indexPattern.getRuntimeMappings();
};
```

Thanks for introducing that shortcut!

mattkime replied:
@weltenwort I implemented your suggestion...at least I think I did...but now some tests are failing and I'm quite confused as to what might be expected. Do you mind taking a look?

weltenwort replied:
Sorry for the delay. I took the liberty of pushing a fix for the tests. The assertion had to be adjusted slightly because we're not manually injecting the `"lang": "painless"` anymore, but instead we're passing through the mappings provided by the data view.

weltenwort reviewed (changes requested):
`infra` changes LGTM except for one function I commented on below. With the new `getRuntimeMappings()` method on the `DataView` our manual reconstruction is not necessary anymore.

mattkime commented:
@elasticmachine merge upstream

stratoula commented:
@mattkime this works fine. Thanx for the script, it helped a lot!
I tested TSVB, vega, timelion and Lens and they work fine.

I only found one weird thing on Lens
<img width="1080" alt="image" src="https://user-images.githubusercontent.com/17003240/154081152-ad8a3651-b5f4-4173-a344-aca5ccc67ecf.png">

This is something that needs to be fixed but this is something that the @elastic/kibana-vis-editors team is going to fix after this PR is merged. So dont worry about it.

qn895 commented:
ML changes (Index data visualizer and transform) LGTM

Dosant commented on src/plugins/data_views/server/rest_api_routes/util/schemas.ts:
```diff
@@ -72,5 +72,27 @@ export const runtimeFieldSpec = {
+  ),
+  format: schema.maybe(serializedFieldFormatSchema),
+  customLabel: schema.maybe(schema.string()),
+  popularity: schema.maybe(
```
I noticed some desync issues with these fields. 

For example:

1. POST runtime_field with `popularity`
2. See `count` is  correct in `fields`, but no `fieldAttrs` info in `data_view` object 


And also: 

1. POST runtime_field with `popularity` 
2. GET data_view - see it in `fieldAttrs` 

```
"fieldAttrs": {
            "composite_runtime.b": {
                "count": 100
            },
            "composite_runtime.a": {}
        },
```

3. DELETE runtime_field
4. GET data_view
5. see `fieldAttrs`
```
"fieldAttrs": {
            "composite_runtime.b": {
                "count": 100
            }
        },
```


I assume we want fieldAttrs cleaned up and not even get desynced?

mattkime replied:
Yes, this needs fixing. Need to add some tests around this.

Dosant reviewed (commented):
LGTM, noticed a couple edge case we probably should address

mattkime commented on src/plugins/data_views/server/rest_api_routes/runtime_fields/create_runtime_field.ts:
```diff
@@ -45,18 +45,15 @@ export const createRuntimeField = async ({
   const dataView = await dataViewsService.get(id);
 
-  if (dataView.fields.getByName(name)) {
+  if (dataView.fields.getByName(name) || dataView.getRuntimeField(name)) {
```
I've added an additional check but there's no way to prevent this from happening with mapped fields. ES has a 'we try to handle everything thrown at us' philosophy so its possible that `composite_runtime` exists and then there's a mapped field `composite_runtime.a`

mattkime commented:
@elasticmachine merge upstream

mattkime commented:
@Dosant

1) I took another look at this and it appears to be working for me. I should have addressed your comment about cleaning up fieldAttrs - we never clean them up. This payload correctly returned fieldAttrs and the value was correctly set in data views management -

```
{
  "name": "composite_runtime",
  "runtimeField": {
    "type": "composite",
    "script": {
      "source": "emit(\"a\",\"a\"); emit(\"b\",\"b\")"
    },
    "fields": {
      "a": {
        "type": "keyword",
        "popularity": 100
      },
      "b": {
        "type": "keyword"
      }
    }
  }
}
```

Dosant commented:
@mattkime, 

> I took another look at this and it appears to be working for me. I should have addressed your comment about cleaning up fieldAttrs - we never clean them up. This payload correctly returned fieldAttrs and the value was correctly set in data views management -

If you say that this is expected that we don't clean up fieldAttrs after a field is deleted then sounds good 


> I'm not sure how to do this. Do you know how? It seems like a case where we want an 'or' statement in our schema.

Maybe simple through oneOf? Then we can also be specific about runtime field type: 
if type is composite then you get one schema, otherwise another schema

If that won't work, then simply runtime checks inside API route handlers? 

```
export const runtimeFieldSpec = schema.oneOf([
  schema.object({
    type: runtimeFieldNonCompositeFieldsSpecTypeSchema,
    script: schema.maybe(
      schema.object({
        source: schema.string(),
      })
    ),
    format: schema.maybe(serializedFieldFormatSchema),
    customLabel: schema.maybe(schema.string()),
    popularity: schema.maybe(
      schema.number({
        min: 0,
      })
    ),
  }),
  schema.object({
    type: runtimeFieldSpecTypeSchema,
    script: schema.maybe(
      schema.object({
        source: schema.string(),
      })
    ),
    fields: schema.maybe(
      schema.recordOf(
        schema.string(),
        schema.object({
          type: runtimeFieldCompositeFieldsTypeSchema,
          format: schema.maybe(serializedFieldFormatSchema),
          customLabel: schema.maybe(schema.string()),
          popularity: schema.maybe(
            schema.number({
              min: 0,
            })
          ),
        })
      )
    ),
  }),
]);
```

Dosant reviewed (approved):
LGTM + previous suggestions

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 125183` or prevent reminders by adding the `backport:skip` label.
